packages/kbn-expandable-flyout/src/hooks/use_url_state.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { useCallback, useEffect, useState } from 'react';
+import { encode, decode, RisonValue } from '@kbn/rison';
+
+interface StateCache<T> {
+  namespaces: Record<string, Record<string, T | undefined>>;
+  timeout: number;
+}
+
+const cache: StateCache<unknown> = {
+  namespaces: {},
+  timeout: 0,
+};
+
+export const useUrlState = <T = unknown>(rootKey: string, key: string) => {
+  if (!cache.namespaces[rootKey]) {
+    cache.namespaces[rootKey] = {};
+  }
+
+  const [internalValue, setInternalValue] = useState<T | undefined>(undefined);
+
+  useEffect(() => {
+    const listener = () => {
+      const searchParams = new URLSearchParams(window.location.search);
+      const param = searchParams.get(rootKey);
+
+      const decodedState = param ? decode(param) : ({} as Record<string, RisonValue>);
+      const decodedValue = (decodedState as Record<string, RisonValue> | undefined)?.[key];
+      cache.namespaces[rootKey][key] = decodedValue;
+      setInternalValue(decodedValue as unknown as T);
+    };
+
+    listener();
+    window.addEventListener('url:update', listener);
+    window.addEventListener('popstate', listener);
+
+    return () => {
+      window.removeEventListener('url:update', listener);
```

Comment:
should also delete the relevant url params on unmount I think

```
      const urlSearchParams = new URLSearchParams(window.location.search);
      urlSearchParams.delete(rootKey);
      const relativeURL = { search: urlSearchParams.toString() };
      history.replace(relativeURL);
```

Reply:
@kqualters-elastic that query param is used across several hooks potentially, so we cannot do automated cleanup